### PR TITLE
Check content before zip/unzip the txhashset

### PR DESCRIPTION
**Pull request by quentinlesceller**
Fix https://github.com/mimblewimble/grin/issues/1171.

- Check the content of the txhashset dir before zip/unzip.
- Add tests
- Rename ```test_txhashset_raw_txs.rs``` → ```test_txhashset.rs```

### CODE REVIEW

**Comment by antiochp** [comment]
Hoping to get #1147 in before we tackle this - then we can revisit the list of files that the `txhashset.zip` should actually contain.

**Comment by quentinlesceller** [comment]
Technically, with https://github.com/mimblewimble/grin/pull/1174/files#diff-12bff24bc654a13261536be1084749a8R31, this method is "future-proof".

**Comment by antiochp** [comment]
With #1147 we will need to support a more dynamic list of files.
The new `pmmr_utxo.bin` file will be rewound and a snapshot written to disk with a block hash as a suffix on the filename (we need to rewind it _before_ we send it, as we cannot rewind it after receiving it before we process full blocks).
So we will need to support passing in a block header (or similar) as part of the txhashset creation process, and the filtering of files will need to be aware of this.

**Comment by quentinlesceller** [comment]
I see I will update accordingly.

**Comment by ignopeverell** [comment]
What's the status on this PR? It'd be really nice to have, is it just missing some sort of wildcard pattern matching? Do we know the exact name the snapshot will have (I think it includes the height)?

**Comment by quentinlesceller** [comment]
Yeah, I'll try to finish this PR ASAP.

**Comment by quentinlesceller** [comment]
@ignopeverell @antiochp, Just updated what do you think?

File: chain/src/txhashset.rs [hunk 1]
```diff
@@ -1072,14 +1071,76 @@ pub fn zip_read(root_dir: String) -> Result<File, Error> {
...
+		return Err(ErrorKind::Other(
+			"Unexpected file(s) found in txhashset folder".to_string(),
+		).into());
+	}
+
+	// Checking now every subdirectory
+	let pmmr_files_expected: HashSet<_> = PMMR_FILES
+		.iter()
+		.cloned()
+		.map(|s| if s.contains(PMMR_FILES[2]) {
```
Comment (antiochp) [inline chain/src/txhashset.rs]: This `PMMR_FILES[2]` is easy to miss - better to just be explicit about the filename itself?

**Comment by quentinlesceller** [comment]
Sorry for the delay catching up asap.

File: chain/src/txhashset.rs [hunk 2]
```diff
@@ -1072,14 +1071,76 @@ pub fn zip_read(root_dir: String) -> Result<File, Error> {
...
+	check_files(&txhashset_path, header)?;
 
 	fs::create_dir_all(txhashset_path.clone())?;
 	zip::decompress(txhashset_data, &txhashset_path)
 		.map_err(|ze| ErrorKind::Other(ze.to_string()).into())
 }
 
+/// Check that the txhashset directory does not contains unexpected files
+fn check_files(txhashset_path: &PathBuf, header: &BlockHeader) -> Result<(), Error> {
+	let subdirectories_expected: HashSet<_> = [OUTPUT_SUBDIR, KERNEL_SUBDIR, RANGE_PROOF_SUBDIR]
```
Comment (hashmap) [inline chain/src/txhashset.rs]: do we need to always create the set? perhaps build it once?

File: chain/src/txhashset.rs [hunk 3]
```diff
@@ -1072,14 +1071,76 @@ pub fn zip_read(root_dir: String) -> Result<File, Error> {
...
+		.iter()
+		.cloned()
+		.map(|s| String::from(s))
+		.collect();
+	let subdirectories_found: HashSet<_> = fs::read_dir(txhashset_path)
+		.unwrap()
```
Comment (hashmap) [inline chain/src/txhashset.rs]: `unwrap` looks risky

File: chain/src/txhashset.rs [hunk 4]
```diff
@@ -1072,14 +1071,76 @@ pub fn zip_read(root_dir: String) -> Result<File, Error> {
...
+
+	// Checking now every subdirectory
+	let pmmr_files_expected: HashSet<_> = PMMR_FILES
+		.iter()
+		.cloned()
+		.map(|s| if s.contains(PMMR_FILES[2]) {
+		format!("{}.{}", s, header.hash())}
+		else {String::from(s) })
+		.collect();
+	let subdirectories = fs::read_dir(txhashset_path).unwrap();
```
Comment (hashmap) [inline chain/src/txhashset.rs]: `unwrap` here and below

File: chain/src/txhashset.rs [hunk 5]
```diff
@@ -1072,14 +1071,76 @@ pub fn zip_read(root_dir: String) -> Result<File, Error> {
...
+	for subdirectory in subdirectories {
+		let subdirectory_path = subdirectory.unwrap().path();
+		let pmmr_files = fs::read_dir(subdirectory_path).unwrap();
```
Comment (hashmap) [inline chain/src/txhashset.rs]: this code looks like a copy code above starting from line 1090, can we make euse it somehow?

File: chain/src/txhashset.rs [hunk 6]
```diff
@@ -1072,14 +1071,76 @@ pub fn zip_read(root_dir: String) -> Result<File, Error> {
...
+			})
+		})
+		.collect();
+	let dir_difference: Vec<String> = subdirectories_found
+		.difference(&subdirectories_expected)
+		.cloned()
+		.collect();
+	if !dir_difference.is_empty() {
+		return Err(ErrorKind::Other(
+			"Unexpected file(s) found in txhashset folder".to_string(),
```
Comment (hashmap) [inline chain/src/txhashset.rs]: Should we care about missing files?

**Comment by quentinlesceller** [comment]
@ignopeverell @antiochp @hashmap finally found time to finish this. Now we filter the content of the txhashset folder on both sides.
When we zip, we create a copy of the txhashset dir, we remove any unexpected files from there and zip it.
When we receive a zip, we unzip it and filter out unexpected files.

**Comment by antiochp** [comment]
:+1: Awesome!

**Pull request merged**